### DISCUSSION TITLE
dpdata doesn't read the CP2K trajectory

### DISCUSSION THREAD

**Discussion by hanao2** [deepmodeling/dpdata#460] [General]
Hello,

I want to convert my CP2K trajectory to DeePMD raw data using dpdata. However, when using t=dpdata.LabeledSystem("output", fmt="cp2k/aimd_output"), I get the following:

```
Data Summary
Labeled System
-------------------
Frame Numbers      : 0
Atom Numbers       : 0
Including Virials  : No
Element List       :
-------------------
```

My trajectory contains 4 frames; therefore, there is some problem going around with my dpdata processing. I have 4 files named "cp2k.log", "IL-1.ener", "IL-frc-1.xyz" and "IL-pos-1.xyz" in the "output" directory and I tried to replicate the necessary keys from dpdata's tests/examples in my input file ("input.txt" which is attached here). The versions of dpdata and cp2k are 0.2.14 and 2022.1, respectively. Could you please help me with fixing the issue?

Thanks!
[input.txt](https://github.com/deepmodeling/dpdata/files/11337972/input.txt)

**Comment by robinzyb** [comment]
try installing the cp2kdata package and use `fmt='cp2kdata/md'`. see https://github.com/deepmodeling/dpdata/issues/439 for detail

**Reply by hanao2** [reply]
Hello,

Thanks for your response. Upon using the 'cp2kdata/md' format, I get the following error:
```
import dpdata
import cp2kdata
t=dpdata.LabeledSystem('output', 'cp2kdata/md')
--- You are parsing data using package Cp2kData ---
Obtian Energies From output/IL-1.ener
Obtian Structures From output/IL-pos-1.xyz
Obtian Froces From output/IL-frc-1.xyz
Missing the atomic kind informations, atom names are true chemical symbols.
No cell information, please check if your inputs are correct.
--- You are parsing data using package Cp2kData ---
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/hana/.local/lib/python3.8/site-packages/dpdata/system.py", line 283, in __init__
    self.from_fmt(
  File "/home/hana/.local/lib/python3.8/site-packages/dpdata/system.py", line 320, in from_fmt
    return self.from_fmt_obj(load_format(fmt), file_name, **kwargs)
  File "/home/hana/.local/lib/python3.8/site-packages/dpdata/system.py", line 1190, in from_fmt_obj
    self.check_data()
  File "/home/hana/.local/lib/python3.8/site-packages/dpdata/system.py", line 307, in check_data
    dd.check(self)
  File "/home/hana/.local/lib/python3.8/site-packages/dpdata/system.py", line 118, in check
    raise DataError(
dpdata.system.DataError: Type of cells is NoneType, but expected ndarray
```
How can I fix it?

Thanks!

**Comment by robinzyb** [comment]
hello,
    dpdata.LabeledSystem('output', 'cp2kdata/md') should be 
```
import dpdata
cp2kmd_dir = "."
cp2kmd_output_name = "output"
dp = dpdata.LabeledSystem(cp2kmd_dir, cp2k_output_name=cp2kmd_output_name, fmt="cp2kdata/md")
print(dp)
```
https://github.com/robinzyb/cp2kdata please check out the readme in this repository

**Reply by hanao2** [reply]
Hello,

Thanks for your reply. I am still getting the same error. I should note that "IL-1.ener", "IL-frc-1.xyz", "IL-pos-1.xyz" and "output" files are in the working directory where "output" is the cp2k log file. I have two question:

1. Does the index of first frame/snapshot matter? My trajectory starts from index 0.
2. I am using an xyz file in my cp2k input script to read the atomic coordinates. Should I add them directly to the input file?

Please let me know your comments.

Thanks!

**Reply by robinzyb** [reply]
hello,
      it should not get same error. Did you replace the cp2kmd_output_name = "cp2k.log"? "output" is my cp2klog name, it should be replaced accordingly.
     for questions:
    1. it doesn't matter
    2. No required. cp2kdata has found it as you see in 
```
Obtian Energies From output/IL-1.ener
Obtian Structures From output/IL-pos-1.xyz
Obtian Froces From output/IL-frc-1.xyz
```

**Reply by robinzyb** [reply]
in your case it should be 
```
cp2kmd_dir = "output"
cp2kmd_output_name = "cp2k.log"
dp = dpdata.LabeledSystem(cp2kmd_dir, cp2k_output_name=cp2kmd_output_name, fmt="cp2kdata/md")
```
The `cp2kdata` will find `output/cp2k.log`, `output/ IL-1.ener`, `output/IL-frc-1.xyz` and `output/IL-pos-1.xyz` accordingly.

**Reply by hanao2** [reply]
I created a directory named "output" and moved the "IL-1.ener", "IL-frc-1.xyz", "IL-pos-1.xyz" and "cp2k.log" files to that. Then, I followed your instruction. Here is what I got:

```
import dpdata
cp2kmd_dir = "output"
cp2kmd_output_name = "cp2k.log"
dp = dpdata.LabeledSystem(cp2kmd_dir, cp2k_output_name=cp2kmd_output_name, fmt="cp2kdata/md")
--- You are parsing data using package Cp2kData ---
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/hananeh/.local/lib/python3.8/site-packages/dpdata/system.py", line 283, in __init__
    self.from_fmt(
  File "/home/hananeh/.local/lib/python3.8/site-packages/dpdata/system.py", line 320, in from_fmt
    return self.from_fmt_obj(load_format(fmt), file_name, **kwargs)
  File "/home/hananeh/.local/lib/python3.8/site-packages/dpdata/system.py", line 1183, in from_fmt_obj
    data = fmtobj.from_labeled_system(file_name, **kwargs)
  File "/home/hananeh/.local/lib/python3.8/site-packages/cp2kdata/dpdata_plugin.py", line 71, in from_labeled_system
    cp2kmd = Cp2kOutput(output_file=cp2k_output_name, run_type="MD", path_prefix=path_prefix)
  File "/home/hananeh/.local/lib/python3.8/site-packages/cp2kdata/output.py", line 83, in __init__
    self.Cp2kInfo = parse_cp2k_info(self.filename)
  File "/home/hananeh/.local/lib/python3.8/site-packages/cp2kdata/block_parser/header_info.py", line 24, in parse_cp2k_info
    return Cp2kInfo(version=cp2k_info["version"][0][0][0])
IndexError: list index out of range
```

**Reply by robinzyb** [reply]
Thank you for the report. It seems the current parser cannot read the cp2k version  2022.1. Can you provide the output file to me. I will update the `cp2kdata` package tomorrow.

**Reply by hanao2** [reply]
Thanks for your help. I am attaching the position, energy, force, stress and log files here.

[cp2k_files.zip](https://github.com/deepmodeling/dpdata/files/11346257/cp2k_files.zip)

**Reply by robinzyb** [reply]
hi, 
   I have updated the cp2kdata, please git clone the devel branch and `pip install .` "https://github.com/robinzyb/cp2kdata/tree/devel"

The following is test results
```python
cp2k_md_dir = "./cp2k_files"
cp2k_output_name = "cp2k.log"
x= dpdata.LabeledSystem(cp2k_md_dir, cp2k_output_name=cp2k_output_name, fmt="cp2kdata/md")
x.to_deepmd_npy("deepmd")
```
```shell
--- You are parsing data using package Cp2kData ---
You are reading cell information from [./cp2k_files/cp2k.log]
Obtian Energies From ./cp2k_files/IL-1.ener
Obtian Structures From ./cp2k_files/IL-pos-1.xyz
Obtian Froces From ./cp2k_files/IL-frc-1.xyz
Obtian Stresses From ./cp2k_files/IL-1.stress
Atom names are fake chemical symbols as you set in cp2k input.
--- You are parsing data using package Cp2kData ---
```

**Reply by hanao2** [reply]
Hello,

Thanks a lot for the update. I could successfully convert my data!

**Reply by hanao2** [reply]
I have one more question. In the dpdata module is there any implementation for converting Wannier centers' data to deepmd raw data?

Thanks!

**Reply by robinzyb** [reply]
@njzjz @amcadmus Do we have any implementation? I have a simple script for cp2k, but I did not find any datatype for `atomic dipole` in dpdata. BTW, what is the usage of `atom_pref` in `DataType("atom_pref", np.ndarray, (Axis.NFRAMES, Axis.NATOMS), required=False)`

**Reply by wanghan-iapcm** [reply]
atomic dipole is not supported by `dpdata` yet. The `atom_pref` is the prefactor multiplied on the force term in the loss function. By `atom_pref` the error of force prediction on some atoms may be paid with more weights.

**Reply by robinzyb** [reply]
How about add new datatype `DataType("atomic_dipole", np.ndarray, (Axis.NFRAMES, <?>), required=False)`. But the <?> should not use `Axis.NATOMS` because it is associated with the part of atoms and the dimension will be less than the `Axis.NATOMS`

**Reply by njzjz** [reply]
It's unclear to me how it will work when calling `sort_atom_types` or concatenating two frames with different shapes of `atomic_dipole`. On another hand, does DeePMD-kit support it?

**Reply by robinzyb** [reply]
1. Currently, the atomic_dipole is strongly associated with `atom_types`.
   For example: 
          1. in pure `H2O` systems, the wannier centroids are identical to the order of the oxygen atoms
          2. a more complex example is `NaCl+H2O` aqueous system. Assuming that the order of chemical symbols is `HO(1)NaHClHO(2)HHO(3)H`(numbers in parenthesis indicate the indices of particle elements.) and the only the `Na`, `Cl`, and `O` have `atomic_dipole` properties. the `atomic_dipole.npy` should  be `O(1)NaClO(2)O(3)`. If we sort the types, the sorted order should be `ClHHHHHHNaO(1)O(2)O(3)` and the data in `atomic_dipole.npy` should have the order of `ClNaO(1)O(2)O(3)`
3. all frames (in the same system) have same shapes which will be less than Natoms.

**Reply by njzjz** [reply]
It seems to me that deepmd-kit still reads data with the shape of natoms but then makes a slice. If so, I prefer storing and dumping data with `Axis.NATOMS`.

https://github.com/deepmodeling/deepmd-kit/blob/45a8d1e3a11f5c4621254ea9552015f724e5593c/deepmd/utils/data.py#L541

**Reply by robinzyb** [reply]
`DeepmdData` resets the natom if `type_sel`  is not `None`.
https://github.com/deepmodeling/deepmd-kit/blob/45a8d1e3a11f5c4621254ea9552015f724e5593c/deepmd/utils/data.py#L537-L540
and See the tensor model for the `type_sel`:
https://github.com/deepmodeling/deepmd-kit/blob/3ac8c4c77b61ee4c645cfb59eaefb104d73c0309/deepmd/loss/tensor.py#L59
and the `type_sel` reads from https://github.com/deepmodeling/deepmd-kit/blob/3ac8c4c77b61ee4c645cfb59eaefb104d73c0309/examples/water/dplr/train/dw.json#L28

It seems to me that if we store the `atomic_dipole` as `natoms=192` for 64 `H2O` to file `atomic_dipole.npy`. and actually the `natoms` reads 64 if we set `sel_type=[0]` in `input.json`. Then this line may cause errors. https://github.com/deepmodeling/deepmd-kit/blob/45a8d1e3a11f5c4621254ea9552015f724e5593c/deepmd/utils/data.py#LL556C1-L556C1

**Reply by njzjz** [reply]
Interesting to know this behavior, which is a bit complicated for dpdata.

**Reply by hanao2** [reply]
Hello all,

Thanks for the ongoing discussion. Are you trying to add this feature (conversion of atomic dipoles) to the package?

Thanks!

**Reply by hanao2** [reply]
Hello,

I managed to write a python script to convert the WC output into molecular dipoles. However, in future development, it would be nice to include this feature in the dpdata package. Thanks for your time and help!

**Reply by robinzyb** [reply]
Hello,
      It is great that you have figured it out.  I will try to include it in the near future!

**Comment by hanao2** [comment]
Hello,

Thanks again for your help with my previous issue on converting CP2K AIMD data. Currently, I want to convert a single frame using the 'cp2k/output' format. However, I get the following error: "Shape of forces is (1, 34, 3), but expected (1, 33, 3)"

Do you think if this is caused by cp2k version inconsistency? If not, how can I solve it? I attach the log file for your notice.

Thank you!



[cp2k.log](https://github.com/deepmodeling/dpdata/files/11899887/cp2k.log)

**Reply by njzjz** [reply]
Should be fixed by https://github.com/deepmodeling/dpdata/pull/466

**Reply by hanao2** [reply]
Thanks! it solved the issue. However, I don't see the virial file at the output. any suggestions on how to get it?

**Reply by robinzyb** [reply]
I tested the log file with cp2kdata. 
```python
x=dpdata.LabeledSystem("./cp2k.log", fmt="cp2kdata/e_f")
x.to_deepmd_npy("cp2k_dpmd")
```
It works and has an virial file

**Reply by robinzyb** [reply]
Can We close this discussion and open a new one, I think this discussion has been mixed with different problems. @njzjz
